proxy-router/internal/morroc/mor_rpc.go

```diff
@@ -0,0 +1,268 @@
+
+	timestamp := m.generateTimestamp()
+	params := map[string]interface{}{
+		"message":   reportStr,
+		"timestamp": timestamp,
+	}
+	signature, err := m.generateSignature(params, providerPrivateKeyHex)
+	if err != nil {
+		return RpcMessage{}, err
+	}
+	params["signature"] = signature
+	return RpcMessage{
+		Method: method,
+		Params: params,
+	}, nil
+}
+
+func (m *MorRpc) generateReport(sessionID string, start uint, end uint, prompts uint, tokens uint, reqs []ReqObject) SessionReport {
+	return SessionReport{
+		SessionID: sessionID,
+		Start:     start,
+		End:       end,
+		Prompts:   prompts,
+		Tokens:    tokens,
+		Reqs:      reqs,
+	}
+}
+
+// User Node Communication
+
+func (m *MorRpc) InitiateSessionRequest(user string, provider string, userPubKey string, spend float64, userPrivateKeyHex string) (RpcMessage, error) {
+	method := "session.request"
+	timestamp := m.generateTimestamp()
+	params := map[string]interface{}{
+		"timestamp": timestamp,
+		"user":      user,
+		"provider":  provider,
+		"key":       userPubKey,
+		"spend":     fmt.Sprintf("%f", spend),
+	}
+
+	signature, err := m.generateSignature(params, userPrivateKeyHex)
+	if err != nil {
+		return RpcMessage{}, err
+	}
+	params["signature"] = signature
+	return RpcMessage{
+		Method: method,
+		Params: params,
+	}, nil
+}
+
+func (m *MorRpc) SessionPromptRequest(sessionID string, prompt string, providerPubKey string, userPrivateKeyHex string) (RpcMessage, error) {
+	method := "session.prompt"
+	timestamp := m.generateTimestamp()
+	params := map[string]interface{}{
+		"message":   prompt,
+		"sessionid": sessionID,
+		"timestamp": timestamp,
+	}
+	signature, err := m.generateSignature(params, userPrivateKeyHex)
+	if err != nil {
+		return RpcMessage{}, err
+	}
+	params["signature"] = signature
+	return RpcMessage{
+		Method: method,
+		Params: params,
+	}, nil
+}
+
+func (m *MorRpc) SessionCloseRequest(sessionID string, userPrivateKeyHex string) (RpcMessage, error) {
+	method := "session.close"
+	timestamp := m.generateTimestamp()
+	params := map[string]interface{}{
+		"sessionid": sessionID,
+		"timestamp": timestamp,
+	}
+	signature, err := m.generateSignature(params, userPrivateKeyHex)
+	if err != nil {
+		return RpcMessage{}, err
+	}
+	params["signature"] = signature
+	return RpcMessage{
+		Method: method,
+		Params: params,
+	}, nil
+}
+
+func (m *MorRpc) generateTimestamp() int64 {
+	now := time.Now()
+	return now.UnixMilli()
+}
+
+// https://goethereumbook.org/signature-generate/
+func (m *MorRpc) generateSignature(params map[string]interface{}, privateKeyHex string) (string, error) {
+	resultStr := ""
+
+	keys := make([]string, 0)
+	for k, _ := range params {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+
+	// Concatenate the parameters in the order of the sorted keys
+	key_values := make([]string, 0)
+	for _, k := range keys {
+		key_values = append(key_values, fmt.Sprintf("%s=%v", k, params[k]))
+	}
+	resultStr = strings.Join(key_values, "&")
```

Comment:
my main concern about this - ordering. JSON encoding does not guarantee the order of keys. On the receiver side, when you built an object from 'params' and serialized that - you might have different keys in order and your signature verification will fail. That's why I use alphabet ordering here

Reply:
Valid point. I think the signature should be calculated from the whole message as is, instead of picking fields to avoid confusion. And using json just fits here perfectly. We can impose additional requirements to the json encoding. Golang serializer keeps order of the struct fields and sorts maps lexicographically. 

Summarizing: 
sender: json.Marshall -> hash -> encrypt -> append as signature field -> send
receiver: json.Unmarshall -> remove signature field -> hash -> decrypt signature from sender -> ensure hashes match

Reply:
done